subsys/net/l2/ethernet/ethernet.c

```diff
@@ -569,6 +571,9 @@ static int ethernet_send(struct net_if *iface, struct net_pkt *pkt)
 #endif
 	ret = net_pkt_get_len(pkt);
 
+	/* remove the L2 header from the packet before returning */
+	net_buf_pull(frag, frag->len);
```

Comment:
The problem with that is it will hold off an unused net_buf until it get unreferenced.

Reply:
I wonder if it is correct to just pull the L2 header, while the ethernet driver still holds a reference of the packet which has just been queued in the DMA engine, and maybe not yet sent. This is the case for the SAM E70 GMAC driver for example. Judging by the comments (https://github.com/zephyrproject-rtos/zephyr/blob/master/drivers/ethernet/eth_mcux.c#L90-L104) the MCUX driver might also do that on the future.

Reply:
what guarantees do you have on the return from send?  If you cannot guarantee it has hit the wire, then I'd say that you have a ticking time bomb there.

Reply:
@tbursztyka we can probavbly find a way to free the buf.  I suppose cloning this might work.

Reply:
@agross-linaro On the return from send you have the guarantee that the pkt has one more reference, so it won't be freed until it has been sent. But I don't think you have much more than that. This allows sending the packet without any memory copy.

Reply:
Just to be clear, the packet is reference when queuing the packet:
https://github.com/zephyrproject-rtos/zephyr/blob/a1a6174b37d0aec82bc7deb9195aeb64b563a352/drivers/ethernet/eth_sam_gmac.c#L1408-L1409

and unreferenced in the Ethernet ISR when it has been sent:
https://github.com/zephyrproject-rtos/zephyr/blob/a1a6174b37d0aec82bc7deb9195aeb64b563a352/drivers/ethernet/eth_sam_gmac.c#L564

Reply:
@aurel32 And the tcp code has a ref as well, or it better, to hold on to that until it is ack'd and safe to dispose of.

Reply:
@aurel32 so then we can't really remove the header because there is a race to send before we remove that l2 header.  Ugh.  There has to be a better way...

Reply:
Would net_buf_clone() be an option? Depending on what kind of backing net_buf pool is used that will only allocate a new net_buf but not the data payload (it instead keeps a ref-count on how many net_buf's are referencing the data)